File: grasshopper-plugin/Messages/Commands.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace rhino_zmq_poc
{
    public class Position
    {
        [JsonPropertyName("x")]
        public double X { get; set; }

        [JsonPropertyName("y")]
        public double Y { get; set; }
    }

    public class AddComponentParams
    {
        [JsonPropertyName("typeGuid")]
        public string TypeGuid { get; set; }

        [JsonPropertyName("position")]
        public Position Position { get; set; }

        [JsonPropertyName("preview")]
        public bool Preview { get; set; } = false;
    }

    public class DeleteComponentParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }
    }

    public class ConnectWireParams
    {
        [JsonPropertyName("from")]
        public PortRef From { get; set; }

        [JsonPropertyName("to")]
        public PortRef To { get; set; }
    }

    public class DisconnectWireParams
    {
        [JsonPropertyName("from")]
        public PortRef From { get; set; }

        [JsonPropertyName("to")]
        public PortRef To { get; set; }
    }

    public class MoveComponentParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }

        [JsonPropertyName("position")]
        public Position Position { get; set; }
    }

    public class RenameComponentParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }

        [JsonPropertyName("nickName")]
        public string NickName { get; set; }
    }

    public class SetComponentLockedParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }

        [JsonPropertyName("locked")]
        public bool Locked { get; set; }
    }

    public class SetComponentHiddenParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }

        [JsonPropertyName("hidden")]
        public bool Hidden { get; set; }
    }

    public class AddGroupParams
    {
        [JsonPropertyName("componentIds")]
        public string[] ComponentIds { get; set; }

        [JsonPropertyName("groupName")]
        public string GroupName { get; set; }

        [JsonPropertyName("color")]
        public string Color { get; set; } = "rgba(255,255,255,150)";

        [JsonPropertyName("border")]
        public string Border { get; set; }
    }

    public class RemoveFromGroupParams
    {
        [JsonPropertyName("componentIds")]
        public string[] ComponentIds { get; set; }

        [JsonPropertyName("groupName")]
        public string GroupName { get; set; }
    }

    public class DeleteGroupParams
    {
        [JsonPropertyName("groupName")]
        public string GroupName { get; set; }
    }

    public class ChangeGroupColorParams
    {
        [JsonPropertyName("groupName")]
        public string GroupName { get; set; }

        [JsonPropertyName("color")]
        public string Color { get; set; } = "rgba(255,255,255,150)";
    }

    public class RenameGroupParams
    {
        [JsonPropertyName("groupName")]
        public string GroupName { get; set; }

        [JsonPropertyName("name")]
        public string Name { get; set; }
    }

    public class ChangeGroupStyleParams
    {
        [JsonPropertyName("groupName")]
        public string GroupName { get; set; }

        [JsonPropertyName("color")]
        public string Color { get; set; }

        [JsonPropertyName("name")]
        public string Name { get; set; }

        [JsonPropertyName("border")]
        public string Border { get; set; }
    }

    public class SetSliderValueParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }

        [JsonPropertyName("value")]
        public double Value { get; set; }
    }

    public class CreateSliderParams
    {
        [JsonPropertyName("position")]
        public Position Position { get; set; }

        [JsonPropertyName("nickName")]
        public string NickName { get; set; }

        [JsonPropertyName("min")]
        public double Min { get; set; }

        [JsonPropertyName("max")]
        public double Max { get; set; }

        [JsonPropertyName("value")]
        public double Value { get; set; }

        [JsonPropertyName("digits")]
        public int Digits { get; set; }

        [JsonPropertyName("interval")]
        public double Interval { get; set; }
    }

    public class EditSliderRangeParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }

        [JsonPropertyName("min")]
        public double Min { get; set; }

        [JsonPropertyName("max")]
        public double Max { get; set; }

        [JsonPropertyName("digits")]
        public int Digits { get; set; }

        [JsonPropertyName("interval")]
        public double Interval { get; set; }
    }

    public class SetPanelTextParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }

        [JsonPropertyName("text")]
        public string Text { get; set; }
    }

    public class CreatePanelParams
    {
        [JsonPropertyName("position")]
        public Position Position { get; set; }

        [JsonPropertyName("nickName")]
        public string NickName { get; set; }

        [JsonPropertyName("text")]
        public string Text { get; set; }

        [JsonPropertyName("width")]
        public double? Width { get; set; }

        [JsonPropertyName("height")]
        public double? Height { get; set; }

        [JsonPropertyName("textOutput")]
        public string TextOutput { get; set; }

        [JsonPropertyName("bgColor")]
        public string BgColor { get; set; }
    }

    public class SetPanelParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }

        [JsonPropertyName("textOutput")]
        public string TextOutput { get; set; }

        [JsonPropertyName("width")]
        public double? Width { get; set; }

        [JsonPropertyName("height")]
        public double? Height { get; set; }

        [JsonPropertyName("bgColor")]
        public string BgColor { get; set; }
    }

    public class CreateToggleParams
    {
        [JsonPropertyName("position")]
        public Position Position { get; set; }

        [JsonPropertyName("nickName")]
        public string NickName { get; set; }

        [JsonPropertyName("value")]
        public bool Value { get; set; }
    }

    public class SetToggleValueParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }

        [JsonPropertyName("value")]
        public bool Value { get; set; }
    }

    public class CreateSwatchParams
    {
        [JsonPropertyName("position")]
        public Position Position { get; set; }

        [JsonPropertyName("nickName")]
        public string NickName { get; set; }

        [JsonPropertyName("color")]
        public string Color { get; set; }
    }

    public class SetSwatchColorParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }

        [JsonPropertyName("color")]
        public string Color { get; set; }
    }

    public class CreateScribbleParams
    {
        [JsonPropertyName("position")]
        public Position Position { get; set; }

        [JsonPropertyName("nickName")]
        public string NickName { get; set; }

        [JsonPropertyName("text")]
        public string Text { get; set; }

        [JsonPropertyName("size")]
        public double? Size { get; set; }
    }

    public class SetScribbleTextParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }

        [JsonPropertyName("text")]
        public string Text { get; set; }
    }

    public class CreateValueListItem
    {
        [JsonPropertyName("name")]
        public string Name { get; set; }

        [JsonPropertyName("value")]
        public string Value { get; set; }
    }

    public class CreateValueListParams
    {
        [JsonPropertyName("position")]
        public Position Position { get; set; }

        [JsonPropertyName("nickName")]
        public string NickName { get; set; }

        [JsonPropertyName("items")]
        public CreateValueListItem[] Items { get; set; }

        [JsonPropertyName("selectedIndex")]
        public int? SelectedIndex { get; set; }
    }

    public class SetValueListSelectedParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }

        [JsonPropertyName("selectedIndex")]
        public int SelectedIndex { get; set; }
    }

    public class ScriptIOParam
    {
        [JsonPropertyName("name")]
        public string Name { get; set; }

        [JsonPropertyName("access")]
        public string Access { get; set; }

        [JsonPropertyName("dataMapping")]
        public string DataMapping { get; set; }

        [JsonPropertyName("simplify")]
        public bool? Simplify { get; set; }

        [JsonPropertyName("reverse")]
        public bool? Reverse { get; set; }

        [JsonPropertyName("typeHint")]
        public string TypeHint { get; set; }

        [JsonPropertyName("previousName")]
        public string PreviousName { get; set; }
    }

    public class CreateScriptNodeParams
    {
        [JsonPropertyName("position")]
        public Position Position { get; set; }

        [JsonPropertyName("language")]
        public string Language { get; set; }

        [JsonPropertyName("code")]
        public string Code { get; set; }

        [JsonPropertyName("nickName")]
        public string NickName { get; set; }

        [JsonPropertyName("inputs")]
        public List<ScriptIOParam> Inputs { get; set; }

        [JsonPropertyName("outputs")]
        public List<ScriptIOParam> Outputs { get; set; }
    }

    public class SetScriptCodeParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }

        [JsonPropertyName("code")]
        public string Code { get; set; }

        [JsonPropertyName("inputs")]
        public List<ScriptIOParam> Inputs { get; set; }

        [JsonPropertyName("outputs")]
        public List<ScriptIOParam> Outputs { get; set; }
    }

    public class SyncScriptParamsParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }

        [JsonPropertyName("inputs")]
        public List<ScriptIOParam> Inputs { get; set; }

        [JsonPropertyName("outputs")]
        public List<ScriptIOParam> Outputs { get; set; }
    }

    public class GetScriptCodeParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }
    }

    public class AddScriptInputParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }

        [JsonPropertyName("name")]
        public string Name { get; set; }

        [JsonPropertyName("access")]
        public string Access { get; set; }

        [JsonPropertyName("dataMapping")]
        public string DataMapping { get; set; }

        [JsonPropertyName("simplify")]
        public bool? Simplify { get; set; }

        [JsonPropertyName("reverse")]
        public bool? Reverse { get; set; }

        [JsonPropertyName("typeHint")]
        public string TypeHint { get; set; }
    }

    public class RemoveScriptInputParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }

        [JsonPropertyName("name")]
        public string Name { get; set; }
    }

    public class AddScriptOutputParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }

        [JsonPropertyName("name")]
        public string Name { get; set; }

        [JsonPropertyName("dataMapping")]
        public string DataMapping { get; set; }

        [JsonPropertyName("simplify")]
        public bool? Simplify { get; set; }

        [JsonPropertyName("reverse")]
        public bool? Reverse { get; set; }

        [JsonPropertyName("typeHint")]
        public string TypeHint { get; set; }
    }

    public class RemoveScriptOutputParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }

        [JsonPropertyName("name")]
        public string Name { get; set; }
    }

    public class ListScriptParamsParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }
    }

    public class EditParamPropsParams
    {
        [JsonPropertyName("targetId")]
        public string TargetId { get; set; }

        [JsonPropertyName("name")]
        public string Name { get; set; }

        [JsonPropertyName("access")]
        public string Access { get; set; }

        [JsonPropertyName("dataMapping")]
        public string DataMapping { get; set; }

        [JsonPropertyName("simplify")]
        public bool? Simplify { get; set; }

        [JsonPropertyName("reverse")]
        public bool? Reverse { get; set; }

        [JsonPropertyName("typeHint")]
        public string TypeHint { get; set; }
    }

    public class PortRef
    {
        [JsonPropertyName("componentId")]
        public string ComponentId { get; set; }

        [JsonPropertyName("port")]
        public string Port { get; set; }
    }

    public class GhCommand
    {
        [JsonPropertyName("action")]
        public string Action { get; set; }

        [JsonPropertyName("params")]
        public JsonElement Params { get; set; }
    }

    public class SubmitJobRequest
    {
        [JsonPropertyName("type")]
        public string Type { get; set; }

        [JsonPropertyName("token")]
        public string Token { get; set; }

        [JsonPropertyName("jobId")]
        public string JobId { get; set; }

        [JsonPropertyName("command")]
        public GhCommand Command { get; set; }
    }

    public class BeginAgentTransactionParams
    {
        [JsonPropertyName("name")]
        public string Name { get; set; }
    }

    public class RunRhinoScriptRequest
    {
        [JsonPropertyName("type")]
        public string Type { get; set; } = "runRhinoScript";

        [JsonPropertyName("mode")]
        public string Mode { get; set; }

        [JsonPropertyName("source")]
        public string Source { get; set; }

        [JsonPropertyName("echo")]
        public bool Echo { get; set; }
    }
}

File: src/types/commands.ts
export type Position = {
	x: number;
	y: number;
};

export type PortRef = {
	componentId: string;
	port: string;
};

export type AddComponentParams = {
	typeGuid: string;
	position: Position;
	preview?: boolean;
};

export type DeleteComponentParams = {
	targetId: string;
};

export type ConnectWireParams = {
	from: PortRef;
	to: PortRef;
};

export type DisconnectWireParams = {
	from: PortRef;
	to: PortRef;
};

export type MoveComponentParams = {
	targetId: string;
	position: Position;
};

export type RenameComponentParams = {
	targetId: string;
	nickName: string;
};

export type SetComponentLockedParams = {
	targetId: string;
	locked: boolean;
};

export type SetComponentHiddenParams = {
	targetId: string;
	hidden: boolean;
};

export type AddGroupParams = {
	componentIds: string[];
	groupName: string;
	color?: string;
	border?: "Box" | "Blob" | "Rectangles";
};

export type RemoveFromGroupParams = {
	componentIds: string[];
	groupName: string;
};

export type DeleteGroupParams = {
	groupName: string;
};

export type ChangeGroupColorParams = {
	groupName: string;
	color?: string;
};

export type RenameGroupParams = {
	groupName: string;
	name: string;
};

export type ChangeGroupStyleParams = {
	groupName: string;
	color?: string;
	name?: string;
	border?: "Box" | "Blob" | "Rectangles";
};

export type SetSliderValueParams = {
	targetId: string;
	value: number;
};

export type CreateSliderParams = {
	position: Position;
	nickName?: string;
	min: number;
	max: number;
	value: number;
	digits: number;
};

export type EditSliderRangeParams = {
	targetId: string;
	min: number;
	max: number;
	digits: number;
};

export type SetPanelTextParams = {
	targetId: string;
	text: string;
};

export type CreateToggleParams = {
	position: Position;
	nickName?: string;
	value: boolean;
};

export type SetToggleValueParams = {
	targetId: string;
	value: boolean;
};

export type CreateSwatchParams = {
	position: Position;
	nickName?: string;
	color: string;
};

export type SetSwatchColorParams = {
	targetId: string;
	color: string;
};

export type CreateScribbleParams = {
	position: Position;
	nickName?: string;
	text: string;
	size?: number;
};

export type SetScribbleTextParams = {
	targetId: string;
	text: string;
};

export type CreateValueListItem = {
	name: string;
	value: string;
};

export type CreateValueListParams = {
	position: Position;
	nickName?: string;
	items: CreateValueListItem[];
	selectedIndex?: number;
};

export type SetValueListSelectedParams = {
	targetId: string;
	selectedIndex: number;
};

export type PanelTextOutput = "singleString" | "oneItemPerLine";

export type CreatePanelParams = {
	position: Position;
	nickName?: string;
	text: string;
	textOutput: PanelTextOutput;
	width?: number;
	height?: number;
	bgColor?: string;
};

export type SetPanelParams = {
	targetId: string;
	textOutput: PanelTextOutput;
	width?: number;
	height?: number;
	bgColor?: string;
};

export type TypeHint = "object" | "double" | "int" | "integer" | "string" | "bool" | "boolean";

export type ScriptIOParam = {
	name: string;
	previousName?: string;
	access?: string;
	dataMapping?: string;
	simplify?: boolean;
	reverse?: boolean;
	typeHint?: TypeHint;
};

export type CreateScriptNodeParams = {
	position: Position;
	language: "python" | "csharp";
	code: string;
	nickName?: string;
	inputs?: ScriptIOParam[];
	outputs?: ScriptIOParam[];
};

export type SetScriptCodeParams = {
	targetId: string;
	code: string;
	inputs?: ScriptIOParam[];
	outputs?: ScriptIOParam[];
};

export type SyncScriptParamsParams = {
	targetId: string;
	inputs?: ScriptIOParam[];
	outputs?: ScriptIOParam[];
};

export type GetScriptCodeParams = {
	targetId: string;
};

export type AddScriptInputParams = {
	targetId: string;
	name: string;
	access?: string;
	dataMapping?: string;
	simplify?: boolean;
	reverse?: boolean;
	typeHint?: TypeHint;
};

export type RemoveScriptInputParams = {
	targetId: string;
	name: string;
};

export type AddScriptOutputParams = {
	targetId: string;
	name: string;
	dataMapping?: string;
	simplify?: boolean;
	reverse?: boolean;
	typeHint?: TypeHint;
};

export type RemoveScriptOutputParams = {
	targetId: string;
	name: string;
};

export type ListScriptParamsParams = {
	targetId: string;
};

export type EditParamPropsParams = {
	targetId: string;
	name: string;
	access?: "item" | "list" | "tree";
	typeHint?: TypeHint;
	dataMapping?: "none" | "flatten" | "graft";
	simplify?: boolean;
	reverse?: boolean;
};

export type BeginAgentTransactionParams = {
	name?: string;
};

export type EmptyParams = Record<string, never>;

export type RhinoObjectQueryParams = {
	selectionOnly?: boolean;
	layer?: string;
	objectType?: string;
};

export type SetParamRhinoGeometryParams = {
	targetId: string;
	mode: "reference" | "internalize";
	rhinoObjectIds?: string[];
	rhinoQuery?: RhinoObjectQueryParams;
};

export type CommandMap = {
	addComponent: AddComponentParams;
	deleteComponent: DeleteComponentParams;
	connectWire: ConnectWireParams;
	disconnectWire: DisconnectWireParams;
	moveComponent: MoveComponentParams;
	renameComponent: RenameComponentParams;
	setComponentLocked: SetComponentLockedParams;
	setComponentHidden: SetComponentHiddenParams;
	addGroup: AddGroupParams;
	removeFromGroup: RemoveFromGroupParams;
	deleteGroup: DeleteGroupParams;
	changeGroupColor: ChangeGroupColorParams;
	renameGroup: RenameGroupParams;
	changeGroupStyle: ChangeGroupStyleParams;
	createSlider: CreateSliderParams;
	editSliderRange: EditSliderRangeParams;
	setSliderValue: SetSliderValueParams;
	createPanel: CreatePanelParams;
	setPanelParams: SetPanelParams;
	setPanelText: SetPanelTextParams;
	createToggle: CreateToggleParams;
	setToggleValue: SetToggleValueParams;
	createSwatch: CreateSwatchParams;
	setSwatchColor: SetSwatchColorParams;
	createScribble: CreateScribbleParams;
	setScribbleText: SetScribbleTextParams;
	createValueList: CreateValueListParams;
	setValueListSelected: SetValueListSelectedParams;
	createScriptNode: CreateScriptNodeParams;
	setScriptCode: SetScriptCodeParams;
	syncScriptParams: SyncScriptParamsParams;
	getScriptCode: GetScriptCodeParams;
	addScriptInput: AddScriptInputParams;
	removeScriptInput: RemoveScriptInputParams;
	addScriptOutput: AddScriptOutputParams;
	removeScriptOutput: RemoveScriptOutputParams;
	listScriptParams: ListScriptParamsParams;
	editParamProps: EditParamPropsParams;
	beginAgentTransaction: BeginAgentTransactionParams;
	commitAgentTransaction: EmptyParams;
	cancelAgentTransaction: EmptyParams;
	beginRhinoAgentTransaction: BeginAgentTransactionParams;
	commitRhinoAgentTransaction: EmptyParams;
	cancelRhinoAgentTransaction: EmptyParams;
	setParamRhinoGeometry: SetParamRhinoGeometryParams;
};

export type CommandAction = keyof CommandMap;

/** Runtime list of all command actions; keep in sync with CommandMap and C# CommandActionRegistry. */
export const COMMAND_ACTIONS = [
	"addComponent",
	"deleteComponent",
	"connectWire",
	"disconnectWire",
	"moveComponent",
	"renameComponent",
	"setComponentLocked",
	"setComponentHidden",
	"addGroup",
	"removeFromGroup",
	"deleteGroup",
	"changeGroupColor",
	"renameGroup",
	"changeGroupStyle",
	"createSlider",
	"editSliderRange",
	"setSliderValue",
	"createPanel",
	"setPanelParams",
	"setPanelText",
	"createToggle",
	"setToggleValue",
	"createSwatch",
	"setSwatchColor",
	"createScribble",
	"setScribbleText",
	"createValueList",
	"setValueListSelected",
	"createScriptNode",
	"setScriptCode",
	"syncScriptParams",
	"getScriptCode",
	"addScriptInput",
	"removeScriptInput",
	"addScriptOutput",
	"removeScriptOutput",
	"listScriptParams",
	"editParamProps",
	"beginAgentTransaction",
	"commitAgentTransaction",
	"cancelAgentTransaction",
	"beginRhinoAgentTransaction",
	"commitRhinoAgentTransaction",
	"cancelRhinoAgentTransaction",
	"setParamRhinoGeometry",
] as const satisfies readonly CommandAction[];
